broker/cloud_functions/lsst/ps_to_gcs/main.py

```diff
@@ -0,0 +1,198 @@
+
+    Args:
+        event: Pub/Sub message data and attributes.
+            `data` field contains the message data in a base64-encoded string.
+            `attributes` field contains the message's custom attributes in a dict.
+
+        context: The Cloud Function's event metadata.
+            It has the following attributes:
+                `event_id`: the Pub/Sub message ID.
+                `timestamp`: the Pub/Sub message publish time.
+                `event_type`: for example: "google.pubsub.topic.publish".
+                `resource`: the resource that emitted the event.
+    """
+    try:
+        upload_bytes_to_bucket(event, context)
+    # this is raised by blob.upload_from_file if the object already exists in the bucket
+    except PreconditionFailed:
+        # we'll simply pass, and the duplicate alert will go no further in our pipeline
+        pass
+
+
+def upload_bytes_to_bucket(event: dict, context: functions_v1.context.Context) -> None:
+    """Uploads the msg data bytes to a GCP storage bucket."""
+
+    alert_bytes = base64.b64decode(event["data"])  # alert packet, bytes
+    attributes = event.get("attributes", {})
+
+    # unpack the alert and read schema ID
+    header_bytes = alert_bytes[:5]
+    schema_id = deserialize_confluent_wire_header(header_bytes)
+
+    # get and load schema
+    sr_client = SchemaRegistryClient({"url": "https://usdf-alert-schemas-dev.slac.stanford.edu"})
+    schema = sr_client.get_schema(schema_id=schema_id)
+    latest_schema = json.loads(schema.schema_str)
+    schema_version = latest_schema["namespace"].split(".")[1]
+    content_bytes = io.BytesIO(alert_bytes[5:])
+
+    # deserialize the alert and create Alert object
+    alert_dict = fastavro.schemaless_reader(content_bytes, latest_schema)
+    filename = generate_alert_filename(
+        {
+            "objectId": alert_dict["diaObject"]["diaObjectId"],
+            "sourceId": alert_dict["diaSource"]["diaSourceId"],
+            "topic": attributes.get("kafka.topic", "no_topic"),
+            "format": "avro",
+        }
+    )
+
+    # get bucket based on the alert's schema version and store the Avro file
+    bucket = BUCKETS.get(schema_version)
+    blob = bucket.blob(filename)
+    blob.metadata = create_file_metadata(alert_dict, context)
+
+    # raise a PreconditionFailed exception if filename already exists in the bucket using "if_generation_match=0"
+    # let it raise. the main function will catch it and then drop the message.
+    blob.upload_from_file(io.BytesIO(alert_bytes), if_generation_match=0)
+
+    # Cloud Storage says this is not a duplicate, so now we publish the broker's main "alerts" stream
+    return publish_outgoing_alert(
+        topic_name=ALERTS_TOPIC.name,
+        message=alert_bytes,
+        attributes={
+            "objectId": str(alert_dict["diaObject"]["diaObjectId"]),
+            "sourceId": str(alert_dict["diaSource"]["diaSourceId"]),
```

Comment:
```suggestion
            "diaObjectId": str(alert_dict["diaObject"]["diaObjectId"]),
            "diaSourceId": str(alert_dict["diaSource"]["diaSourceId"]),
```

Important not to change the field names.